cli/src/org/partiql/shell/ShellHighlighter.kt

```diff
@@ -148,97 +132,101 @@ internal class ShellHighlighter() : Highlighter {
-        return lineIndexes[lineNumber]?.plus(column)
-    }
+    class OffendingSymbolException(val offendingSymbol: org.antlr.v4.runtime.Token) : Exception()
+}
 
-    /**
-     * Sets the color and thickness of the string based on the [token] type
-     */
-    private fun getStyle(token: Token): AttributedStyle {
-        var style = AttributedStyle()
-        val attrCode = when (token.type) {
-            TokenType.KEYWORD -> if (token.isDataType) AttributedStyle.GREEN else AttributedStyle.CYAN
-            TokenType.AS, TokenType.FOR, TokenType.ASC, TokenType.LAST, TokenType.DESC,
-            TokenType.BY, TokenType.FIRST -> AttributedStyle.CYAN
-            TokenType.LITERAL -> if (token.value is IonString) AttributedStyle.YELLOW else AttributedStyle.BLUE
-            TokenType.ION_LITERAL -> AttributedStyle.YELLOW
-            TokenType.OPERATOR -> AttributedStyle.WHITE
-            TokenType.QUOTED_IDENTIFIER -> AttributedStyle.BRIGHT
-            TokenType.IDENTIFIER -> AttributedStyle.BRIGHT
-            TokenType.MISSING -> AttributedStyle.BLUE
-            TokenType.NULL -> AttributedStyle.BLUE
-            else -> AttributedStyle.WHITE
-        }
-        style = style.foreground(attrCode)
+private fun getTokenStream(input: String): CommonTokenStream {
+    val inputStream = CharStreams.fromStream(input.byteInputStream(StandardCharsets.UTF_8), StandardCharsets.UTF_8)
+    val tokenizer = PartiQLTokens(inputStream)
+    tokenizer.removeErrorListeners()
+    return CommonTokenStream(tokenizer)
+}
 
-        return when (token.type) {
-            TokenType.IDENTIFIER, TokenType.OPERATOR -> style.bold()
-            else -> style
-        }
-    }
+private fun isKeyword(type: Int, text: String): Boolean = PartiQLTokens.VOCABULARY.getSymbolicName(type).equals(text, ignoreCase = true)
+
+private fun isDatatype(type: Int) = when (type) {
+    PartiQLTokens.SMALLINT,
+    PartiQLTokens.INT,
+    PartiQLTokens.INT2,
+    PartiQLTokens.INTEGER,
+    PartiQLTokens.INTEGER2,
+    PartiQLTokens.INT4,
+    PartiQLTokens.INTEGER4,
+    PartiQLTokens.INT8,
+    PartiQLTokens.INTEGER8,
+    PartiQLTokens.BIGINT,
+    PartiQLTokens.REAL,
+    PartiQLTokens.TIMESTAMP,
+    PartiQLTokens.DATE,
+    PartiQLTokens.SYMBOL,
+    PartiQLTokens.STRING,
+    PartiQLTokens.BLOB,
+    PartiQLTokens.CLOB,
+    PartiQLTokens.STRUCT,
+    PartiQLTokens.TUPLE,
+    PartiQLTokens.BAG,
+    PartiQLTokens.LIST,
+    PartiQLTokens.SEXP,
+    PartiQLTokens.DECIMAL,
+    PartiQLTokens.FLOAT,
+    PartiQLTokens.CHAR,
+    PartiQLTokens.CHARACTER,
+    PartiQLTokens.VARYING,
+    PartiQLTokens.VARCHAR,
+    PartiQLTokens.NULL,
+    PartiQLTokens.MISSING,
+    PartiQLTokens.BOOL,
+    PartiQLTokens.BOOLEAN,
+    PartiQLTokens.ANY -> true
+    else -> false
 }
 
-private fun ansi(string: String, style: AttributedStyle) = AttributedString(string, style).toAnsi()
+private fun isIonMode(type: Int, text: String) = type == PartiQLTokens.EOF && text.contains("<EOF>").not()
```

Comment:
Not sure I follow the logic here?

Reply:
Leaving a note there. Essentially, ANTLR treats island modes differently.